afpacket/src/sockets.rs

```diff
@@ -0,0 +1,181 @@
+/// At this time, it's not particularly useful.
+pub struct Addr {
+    inner: libc::sockaddr_storage,
+    len: libc::socklen_t,
+}
+
+/// Represents an unbound `AF_PACKET` socket.  At this phase of a socket's lifecycle, it can be
+/// configured.
+pub struct Socket {
+    fd: libc::c_int,
+}
+
+/// Represents a bound `AF_PACKET` socket. At this phase of a socket's lifecycle, it can be read
+/// to/written from.
+pub struct BoundSocket {
+    fd: libc::c_int,
+    send_addr: libc::sockaddr_ll,
+}
+
+impl Socket {
+    /// Creates a new unbound socket.
+    pub fn new() -> io::Result<Self> {
+        // This block is unsafe because it invokes a syscall. It is safe because it checks the
+        // syscall return value and returns an Err if the sycall fails. It does not interact with
+        // Rust-owned memory, so it's safe as per the borrow checker's rules.
+        let fd = unsafe {
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#socket
+            // man 7 packet
+            let fd = libc::socket(libc::AF_PACKET, libc::SOCK_RAW, libc::ETH_P_ALL.to_be());
+            if fd < 0 {
+                return Err(io::Error::last_os_error())
+            }
+            fd
+        };
+        Ok(Self { fd })
+    }
+
+    /// Binds the socket to a network interface. This function consumes the `Socket` instance, as
+    /// no more configuration options may be safely changed.
+    pub fn bind(self, iface: impl AsRef<CStr>) -> io::Result<BoundSocket> {
+        // This block is unsafe because it invokes syscalls. It is safe because the syscall return
+        // values are checked at each syscall and returned as an Err value. This block should be
+        // tested with miri to confirm that the use of the &CStr it takes as input is correct.
+        let send_addr = unsafe {
+            // get the index of the interface
+            let mut ifr : linux::ifreq = MaybeUninit::zeroed().assume_init();
+            ptr::copy_nonoverlapping(iface.as_ref().as_ptr(),
+                ifr.ifr_ifrn.ifrn_name.as_mut_ptr(), libc::IFNAMSIZ);
+            // ioctl(SIOCGIFINDEX) fills in the index field of the ifreq object
+            // Resources:
+            // man 7 netdevice
+            let err = libc::ioctl(self.fd, linux::SIOCGIFINDEX, &ifr);
+            if err < 0 {
+                return Err(io::Error::last_os_error())
+            }
+
+            // bind the socket
+            let mut ll : libc::sockaddr_ll = MaybeUninit::zeroed().assume_init();
+            ll.sll_family = libc::AF_PACKET as libc::c_ushort;
+            ll.sll_ifindex = ifr.ifr_ifru.ifru_ivalue; // expanded from `ifr_ifindex` in kernel headers
+            // Resources:
+            // https://beej.us/guide/bgnet/html/multi/syscalls.html#bind
+            // man 7 packet regarding sockaddr_ll
+            let err = libc::bind(self.fd,
+                                 &mut ll as *mut _ as *mut libc::sockaddr,
```

Comment:
`as *mut _ as *mut libc::sockaddr` excuse me wat?

Reply:
The first part `&mut ll as *mut _` converts `&mut ll` from `&mut libc::sockaddr_ll` to `*mut libc::sockaddr_ll`. After we've converted it to a raw pointer, we then cast the raw pointer to the right type for the function (`*mut libc::sockaddr`).

Reply:
I guess I've never seen a reference casted to a raw pointer before. Strange that we can't do `&mut ll as *mut libc::sockaddr`.

Reply:
Just wanted to follow up that only the following casts are valid (explaining why we need two casts):
* `&T as *const T`
* `&mut T as *mut T`
* `*const T as *const U`
* `*mut T as *mut U`
* `*mut T as *const T`